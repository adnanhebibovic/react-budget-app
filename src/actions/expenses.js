import uuid from 'uuid'

export const addExpense = function({title = '', amount = 0} = {}) {
    return {
        type: 'ADD_EXPENSE',
        expense: {
            id: uuid(),
            title,
            amount,
            createdAt: new Date().getTime()
        }
    }
}

export const removeExpense = function(id = uuid()) {
    return {
        type: 'REMOVE_EXPENSE',
        id
    }
}

export const editExpense = function(id = uuid(), expense = {}) {
    return {
        type: 'EDIT_EXPENSE',
        id,
        expense
    }
}