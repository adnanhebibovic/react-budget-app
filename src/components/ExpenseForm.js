import React, { useState } from 'react'

export default function(props) {
    const [title, setTitle] = useState(props.expense ? props.expense.title : '')
    const [amount, setAmount] = useState(props.expense ? props.expense.amount : '')
    const [error, setError] = useState('')

    function onFormSubmit(e) {
        e.preventDefault();

        if (!(title || amount)) {
            setError('Please provide title and amount!')
        } else {
            props.onSubmit({
                title: title,
                amount: parseFloat(amount)
            })
        }
    }

    function onTextChange(e) {
        const title =  e.target.value;
        setTitle(title);
    }

    function onAmountChange(e) {
        const amount = e.target.value;
        setAmount(amount);
    }

    return (
        <div>
            {error && <p>{error}</p>}
            <form onSubmit={onFormSubmit}>
                <input 
                    type="text" 
                    placeholder="Title" 
                    autoFocus
                    value={title}
                    onChange={onTextChange}></input>
                    
                <input 
                    type="number" 
                    placeholder="Amount"
                    value={amount}
                    onChange={onAmountChange}></input>

                <input type="submit" value="Save"></input>
            </form>
        </div>
    )    
}
