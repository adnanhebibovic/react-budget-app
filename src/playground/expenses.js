import { createStore, combineReducers } from 'redux'
import uuid from 'uuid'

const addExpense = function({title = '', amount = 0} = {}) {
    return {
        type: 'ADD_EXPENSE',
        expense: {
            id: uuid(),
            title,
            amount,
            createdAt: new Date()
        }
    }
}

const removeExpense = function(id = uuid()) {
    return {
        type: 'REMOVE_EXPENSE',
        id
    }
}

const editExpense = function(id = uuid(), expense = {}) {
    return {
        type: 'EDIT_EXPENSE',
        id,
        expense
    }
}

const expensesDefaultState = []

const expensesReducer = function(state = expensesDefaultState, action) {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return state.concat(action.expense)
        case 'REMOVE_EXPENSE':
            return state.filter((expense) => expense.id !== action.id)
        case 'EDIT_EXPENSE':
            return state.map((expense) => expense.id == action.id 
                ? {...expense, ...action.expense}
                : expense
            )
        default:
            return state;
    }
}

const setTextFilter = function(text = undefined) {
    return {
        type: 'SET_TEXT_FILTER',
        text
    }
}

const sortByAmount = function() {
    return {
        type: 'SORT_BY',
        sortBy: 'amount'
    }
}

const sortByDate = function() {
    return {
        type: 'SORT_BY',
        sortBy: 'createdAt'
    }
}

const setDateFilters = function({startDate = undefined, endDate = undefined} = {}) {
    return {
        type: 'SET_DATE_FILTERS',
        dates: {
            startDate,
            endDate
        }
    }
}

const filtersDefaultState = {
    text: undefined,
    sortBy: undefined,
    startDate: undefined,
    endDate: undefined
}

const filtersReducer = function(state = filtersDefaultState, action) {
    switch (action.type) {
        case 'SET_TEXT_FILTER':
            return {
                ...state,
                text: action.text
            }
        case 'SORT_BY':
            return {
                ...state,
                sortBy: action.sortBy
            }
        case 'SET_DATE_FILTERS':
            return {
                ...state,
                ...action.dates
            }
        default:
            return state;
    }
}

const getExpenses = function(expenses, {text, sortBy, startDate, endDate}) {
    const result = expenses.filter((expense) => {
        const textMatch = typeof text === 'undefined' || expense.title.includes(text);
        const startDateMatch = typeof startDate === 'undefined' || expense.createdAt >= startDate;
        const endDateMatch = typeof endDate === 'undefined' || expense.createdAt <= endDate;

        return textMatch && startDateMatch && endDateMatch;
    })

    if (sortBy)
    {
        return result.sort((e1, e2) => {
            if (e1[sortBy] < e2[sortBy]) {
                return -1;
            } else if (e1[sortBy] > e2[sortBy]) {
                return 1;
            } else {
                return 0;
            }
        })
    }

    return result
}

const store = createStore(
    combineReducers({
        expenses: expensesReducer,
        filters: filtersReducer
    })
)

store.dispatch(addExpense({
    title: 'January bills',
    amount: 279.00 
}))

store.dispatch(addExpense({
    title: 'February bills',
    amount: 523.00 
}))

store.dispatch(addExpense({
    title: 'March bills',
    amount: 358.00 
}))

store.dispatch(addExpense({
    title: 'New car',
    amount: 1000.00 
}))

store.subscribe(() => {
    const state = store.getState();
    console.log(getExpenses(state.expenses, state.filters))
})

store.dispatch(setTextFilter('bills'))
store.dispatch(sortByAmount())
store.dispatch(sortByDate())
store.dispatch(setDateFilters({endDate: new Date()}))
