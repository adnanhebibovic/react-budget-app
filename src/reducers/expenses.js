export const expensesDefaultState = []

export default function(state = expensesDefaultState, action) {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return state.concat(action.expense)
        case 'REMOVE_EXPENSE':
            return state.filter((expense) => expense.id !== action.id)
        case 'EDIT_EXPENSE':
            return state.map((expense) => expense.id == action.id 
                ? {...expense, ...action.expense}
                : expense
            )
        default:
            return state;
    }
}