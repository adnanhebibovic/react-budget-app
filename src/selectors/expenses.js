export default function(expenses, {text, sortBy, startDate, endDate}) {
    const result = expenses.filter((expense) => {
        const textMatch = typeof text == 'undefined' || expense.title.toLowerCase().includes(text.toLowerCase());
        const startDateMatch = typeof startDate == 'undefined' || expense.createdAt > startDate;
        const endDateMatch = typeof endDate == 'undefined' || expense.createdAt < endDate;

        return textMatch && startDateMatch && endDateMatch;
    })

    if (sortBy) {
        return result.sort((e1, e2) => {
            if (e1[sortBy] < e2[sortBy]) {
                return -1;
            } else if (e1[sortBy] > e2[sortBy]) {
                return 1;
            } else {
                return 0;
            }
        })
    }

    return result;
}
