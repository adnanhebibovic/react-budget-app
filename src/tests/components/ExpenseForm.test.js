import React from 'react'
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";

import expenses from '../fixtures/expenses';
import ExpenseForm from '../../components/ExpenseForm';

let container = null;

beforeEach(() => {
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

test('Should render Expense form component', () => {
    act(() => {
        render(<ExpenseForm />, container);
    });
    expect(container).toMatchSnapshot()
})

test('Should render Expense form component with expense data', () => {
    act(() => {
        render(<ExpenseForm expense={expenses[0]}></ExpenseForm>, container)
    });
    expect(container).toMatchSnapshot()
})

test('Should render error for invalid ExpenseForm submission', () => {
    const onSubmit = jest.fn()
    act(() => {
        render(<ExpenseForm onSubmit={onSubmit}></ExpenseForm>, container)
    });
    const input = container.querySelector('input[type="submit"]');
    act(() => {
        input.dispatchEvent(new MouseEvent("click", { bubbles: true }));
    });
    const label = container.querySelector('p');
    expect(label.textContent).toBe('Please provide title and amount!')
    expect(onSubmit).not.toHaveBeenCalled()
})

test('Should not render error for valid ExpenseForm submission', () => {
    const onSubmit = jest.fn()
    act(() => {
        render(<ExpenseForm expense={expenses[0]} onSubmit={onSubmit}></ExpenseForm>, container)
    });
    const input = container.querySelector('input[type="submit"]');
    act(() => {
        input.dispatchEvent(new MouseEvent("click", { bubbles: true }));
    });
    const label = container.querySelector('p');
    expect(label).toBe(null)
    expect(onSubmit).toHaveBeenCalledWith({
        title: expenses[0].title,
        amount: expenses[0].amount
    })
})