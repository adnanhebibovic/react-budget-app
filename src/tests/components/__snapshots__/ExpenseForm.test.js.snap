// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Expense form component 1`] = `
<div>
  <div>
    
    <form>
      <input
        placeholder="Title"
        type="text"
        value=""
      />
      <input
        placeholder="Amount"
        type="number"
        value=""
      />
      <input
        type="submit"
        value="Save"
      />
    </form>
  </div>
</div>
`;

exports[`Should render Expense form component with expense data 1`] = `
<div>
  <div>
    
    <form>
      <input
        placeholder="Title"
        type="text"
        value="first expense"
      />
      <input
        placeholder="Amount"
        type="number"
        value="400"
      />
      <input
        type="submit"
        value="Save"
      />
    </form>
  </div>
</div>
`;
